### PR TITLE
feat: allow users in owners file outside of org to trigger pipelines

### LINKED ISSUE #1313: Allow Pull Requests to be automatically tested when the author is in the OWNERS file
Currently, PRs are only automatically tested when an author is in the organization or if it's a Jenkins-X bot. 

My use case would be to allow GitHub's depandabot or other bots PR's to be automatically tested, provided they are added to the OWNERS file, as commenting `/ok-to-test` for each of its PR's is proving to be a bit of a hassle.

### PR DESCRIPTION
**Pull request by ankitm123**
Signed-off-by: ankitm123 <ankitmohapatra123@gmail.com>

fixes #1313

### CODE REVIEW

File: pkg/plugins/trigger/pull-request_test.go [hunk 1]
```diff
@@ -25,17 +25,78 @@ import (
 	"github.com/jenkins-x/lighthouse/pkg/labels"
 	"github.com/jenkins-x/lighthouse/pkg/launcher/fake"
 	"github.com/jenkins-x/lighthouse/pkg/plugins"
+	"github.com/jenkins-x/lighthouse/pkg/repoowners"
 	fake2 "github.com/jenkins-x/lighthouse/pkg/scmprovider/fake"
 	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/sets"
 )
 
+type fakeOwnersClient struct{}
```
Comment (ankitm123) [inline pkg/plugins/trigger/pull-request_test.go:34]: ToDo: This should be in some test helper package as it's now being reused across tests in separate packages.

**Comment by rawlingsj** [comment]
I'm don't think this is a good idea TBH, the aim for the automatic triggering of pipelines come from trusted users which should be managed by the organisation level permissions.  This was a key security design that Prow uses and deviating from that will introduce risk.  

I.e. if a person leaves a company their permissions are managed at the git provider organisation level.  Cleaning up OWNERS files requires updates to all related repos and things can be missed. 

It's important to distinguish between trusted users that can automatically trigger builds, and ultimately edit pipelines that get executed without review which is a significant security risk.  That is different from being in the OWNERS file as an approver of changes.

Let's leave the security model around trusted users and auto triggering pipelines as is.  This is a proven model and works across all Kubernetes repos, any change to this is a security risk.

**Comment by rawlingsj** [comment]
This also means that anyone already in the OWNERS file could approve additional people that would be able to automatically trigger pipelines (and therefore change the behaviour), essentially bypassing the security model currently in place.

Closing this PR due to the security risks in introduces.

**Comment by ankitm123** [comment]
> It's important to distinguish between trusted users that can automatically trigger builds, and ultimately edit pipelines that get executed without review which is a significant security risk

Thinking about it, this may make it less risky: If the user/bot is in OWNERS file and not in the github trusted org, then they cannot change the pipeline files (anything in .lighthouse folder basically) and OWNERS* file. We are detecting which files are being changed. Also this entire feature is turned off by default.

**Comment by rawlingsj** [comment]
Someone could still write a unit test that would be executed without prior approval which did something malicious or change + approve code that should not be done once they have been removed from the org.  I think any change around the prow designed security model is not great TBH as we could open up for exploitation in a scenario like above or one that is not known yet.

**Comment by ankitm123** [comment]
I think the specific use case for the issue addressed by this PR is dependency updates (go mod, package.json gem files etc ...), may be we only trigger for those PRs (so changes outside those files are not going to get triggered), so if a malicious actor opens a PR which changes code (for example tests) other than dependency updates, then it wont trigger (and we can add a comment in the PR explaning to the end user why it was not triggered).

Again by default, all of this is turned off.

The thing that worries me a bit is undesirable third party updates (for example, what happened to faker and color js).
It would be nice if certain bots/apps like renovate/action bot could be added to an org, but I guess that has more security risks ... Let me also ask around in prow channel to see what prow users think ...

**Comment by m1pl** [comment]
After some brainstorming with @ankitm123 in Slack, this is what we gathered so far:
- Bot accounts are distinguishable by their `[bot]` suffix. No regular account can have that.
- It would be good to not hardcode a list of bots, but rather make it configurable (e.g. trustedBots array ["renovate", "dependabot"]).
- I'd rather not limit it to dependency files, maybe just blacklist OWNERS. But even this goes pretty far IMHO. The bots have access to the entire source code at least. If they're not trusted, they wouldn't be there in the first place. So artificially preventing users from doing stuff in jx that would otherwise be no problem is probably not the best idea. As long as we can distinguish bots from regular accounts, I don't see an issue. An organization admin needs to trust the bots, so they can open PRs.
- The PRs shouldn't need the `ok-to-test` label, other things like automerge, e.t.c. should be discussed separately if needed.

**Comment by ankitm123** [comment]
For reference, this is the slack thread: https://kubernetes.slack.com/archives/C9MBGQJRH/p1639664112187200

**Comment by ankitm123** [comment]
I think the source of the confusion was that I did not make it clear that the feature is for github apps (so let's only look at github for now), and not normal users. 
Also the code does not do such checks/distinction atm :facepalm: (the commit message also says users instead of github bots/apps).

**Comment by m1pl** [comment]
Any progress here? This makes it a pain using jx with tools like Renovate or Dependabot.

**Comment by joshuasimon-taulia** [comment]
I'm running self-hosted renovatebot using github app credentials and am unable to automatically test dependency PRs. I have tried configuring renovate to add the `updatebot` label to all PRs

```json
 "packageRules": [
    {
      "description": "Automatically merge minor and patch-level updates",
      "excludePackageNames": [
        "base",
        "gateway",
        "istiod"
      ],
      "matchUpdateTypes": [
        "minor",
        "patch",
        "digest"
      ],
      "addLabels": [
        "updatebot"
      ]
    }
  ]
```

**Pull request closed without merging**